### PR TITLE
feat: Add applicationset metrics

### PR DESCRIPTION
**Pull request by ilia-medvedev**
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: applicationset/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -88,6 +88,7 @@ type ApplicationSetReconciler struct {
 	SCMRootCAPath              string
 	GlobalPreservedAnnotations []string
 	GlobalPreservedLabels      []string
+	Metrics                    metrics.ApplicationsetMetrics
```
Comment (pasha-codefresh) [inline applicationset/controllers/applicationset_controller.go:91]: use pointer

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

**Comment by pasha-codefresh** [comment]
@ilia-medvedev-codefresh could you please execute basic performance tests and make sure that here no performance impact on controller

File: applicationset/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			Name: "argocd_appset_reconcile",
+			Help: "Application reconciliation performance in seconds.",
+			// Buckets can be set later on after observing median time
+		},
+		descAppsetDefaultLabels,
+	)
+
+	appsetCollector := newAppsetCollector(appsetLister, appsetLabels, appsetFilter)
+
+	// Rgister collectors and metrics
```
Comment (pasha-codefresh) [inline applicationset/metrics/metrics.go:56]: Typo

File: applicationset/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	if len(labels) > 0 {
+		descAppsetLabels = prometheus.NewDesc(
+			"argocd_appset_labels",
+			"Applicationset labels translated to Prometheus labels",
+			append(descAppsetDefaultLabels, metricsutil.NormalizeLabels("label", labels)...),
+			nil,
+		)
+	}
+
+	return (&appsetCollector{
```
Comment (pasha-codefresh) [inline applicationset/metrics/metrics.go:81]: Unwrap it ( also run linter, should be fixed by it )

File: applicationset/utils/applicationsetLister.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
+package utils
```
Comment (pasha-codefresh) [inline applicationset/utils/applicationsetLister.go:1]: lets rename the file to use same pattern, applicationset_lister.go

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: util/metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// Prometheus invalid labels, more info: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels.
+var invalidPromLabelChars = regexp.MustCompile(`[^a-zA-Z0-9_]`)
+
+func NormalizeLabels(prefix string, labels []string) []string {
```
Comment (pasha-codefresh) [inline util/metrics/metrics.go:11]: because you already moved it, can you cover this util with tests ?

**Comment by pasha-codefresh** [comment]
@ilia-medvedev-codefresh it also would be nice to have documentation and grafana.json with dashboard

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: util/metrics/metrics.go (hunk 5 above)
Comment (ilia-medvedev-codefresh) [reply util/metrics/metrics.go:11]: Add unit test

**Comment by ilia-medvedev-codefresh** [comment]
I discussed with @kostis-codefresh and he will add metrics documentation in a separate PR

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

**Review by pasha-codefresh — APPROVED** [review]
LGTM, please open PR with documentation and grafana dashboard example

**Pull request merged**